soufi/finders/yum.py

```diff
@@ -0,0 +1,215 @@
+                continue
+            for package in repo.findall(name):
+                # If we have a binary package matching our version, but with
+                # a different name than the corresponding source package,
+                # return the NVR fields
+                if self.version in (package.evr, package.vr):
+                    return self._nevra_or_none(package)
+                # Otherwise let's make it weird
+                packages.append(package)
+
+        # If we've made it here, things have gotten weird; this should be
+        # the case of a source RPM that produces binary RPMs with different
+        # names *and* versions (the lvm2 package from Red Hat is a good example
+        # of this).  In this case, we won't be able to make heads or tails of
+        # the response, unless (and only unless) it contains a single package.
+        try:
+            [package] = packages
+        except ValueError:
+            return None, None
+        return self._nevra_or_none(package)
+
+    def _nevra_or_none(self, package):
+        if package.sourcerpm == '':
+            # It's here, but has no sources defined!  Bummer...
+            return None, None
+        nevra = self._get_nevra(package.sourcerpm)
+        return nevra['name'], f"{nevra['ver']}-{nevra['rel']}"
+
+    # Cache repo downloads as they are slow and network-bound.
+    @classmethod
+    @functools.lru_cache(maxsize=1024)
+    def _get_repo(cls, url):
+        if not url.endswith('/'):
+            url += '/'
+        try:
+            return repomd.load(url)
+        except urllib.error.HTTPError:
+            return None
+
+    # TODO(nic): throw this out and use hawkey/libdnf whenever that finally
+    #  stabilizes.  See: https://github.com/juledwar/soufi/issues/13
+    @staticmethod
+    def _get_nevra(filename):
+        """Split out the NEVRA fields from an RPM filename."""
+
+        # It's easiest to do this by eating the filename backwards, dropping
+        # offset pointers as we go
+        if filename.endswith('.rpm'):
+            filename = filename[:-4]
+        arch_offset = filename.rfind('.')
+        arch = filename[1 + arch_offset :]
+        rel_offset = filename[:arch_offset].rfind('-')
+        rel = filename[1 + rel_offset : arch_offset]
+        ver_offset = filename[:rel_offset].rfind('-')
+        ver = filename[1 + ver_offset : rel_offset]
+        name = filename[:ver_offset]
+        # Sometimes the epoch is before the name, sometimes it's before the
+        # version.  Support both.
+        if ':' in ver:
+            epoch, ver = ver.split(':', maxsplit=1)
+        elif ':' in name:
+            epoch, name = name.split(':', maxsplit=1)
+        else:
+            epoch = ''
+        return dict(name=name, ver=ver, rel=rel, epoch=epoch, arch=arch)
+
+    # Use this wrapper for doing HTTP HEAD requests, as it will swallow
+    # Timeout exceptions, but cache other lookups.
+    @staticmethod
+    def _test_url(url):
```

Comment:
Make this public, it's used by derived classes. Also, a classmethod is appropriate since you're calling other functions defined in the class.